File: lib/javascript/src/math/Vector3D.js

ae.math.Vector3D = class Vector3D {
	
	backend:  ae.math.VectorBackend;
	readOnly: ae.math.Vector3D;
	
	x: number;
	y: number;
	z: number;
	
	constructor(backend: ae.math.VectorBackend) {
		
		this.backend  = backend;
		this.readOnly = backend instanceof ae.math.ReadOnlyBackend ?
			this : new Vector3D(new ae.math.ReadOnlyBackend(backend));
		
		this.x = 0;
		this.y = 0;
		this.z = 0;
	}
	
	add(v: ae.math.Vevtor3D): ae.math.Vevtor3D {
		
		this.backend.x += v.backend.x;
		this.backend.y += v.backend.y;
		this.backend.z += v.backend.z;
		
		return this;
	}
	
	cloneConst(): ae.math.Vevtor3D {
		
		return new Vector3D(new ae.math.ReadOnlyBackend(
			new ae.math.StaticBackend(this.backend)));
	}
	
	cloneDeep(): ae.math.Vevtor3D {
		
		return new Vector3D(this.backend);
	}
	
	cloneStatic(): ae.math.Vevtor3D {
		
		return new Vector3D(new ae.math.StaticBackend(this.backend));
	}
	
	computeLength(): number {
		
		this.copyStaticValues();
		
		return Math.sqrt(this.x * this.x + this.y * this.y + this.z * this.z);
	}

	computeMean(): number {
		
		return (this.backend.x + this.backend.y + this.backend.z) / 3;
	}

	copyStaticValues(): ae.math.Vevtor3D {
		
		this.x = this.backend.x;
		this.y = this.backend.y;
		this.z = this.backend.z;
		
		return this;
	}

	static createConst(
			xOrGrey:  number,
			y:       ?number,
			z:       ?number): ae.math.Vector3D {
		
		y = y || xOrGrey;
		z = z || xOrGrey;
		
		return new Vector3D(new ae.math.ReadOnlyBackend(
			new ae.math.StaticBackend(xOrGrey, y, z)));
	}

	static createStatic(
			xOrGrey:  number,
			y:       ?number,
			z:       ?number): ae.math.Vector3D {
		
		y = y || xOrGrey;
		z = z || xOrGrey;
		
		return new Vector3D(new ae.math.StaticBackend(xOrGrey, y, z));
	}
	
	static cross(
			v1:  ae.math.Vector3D,
			v2:  ae.math.Vector3D,
			dst: ae.math.Vector3D): ae.math.Vector3D {
		
		v1.copyStaticValues();
		v2.copyStaticValues();
		
		dst.backend.x = v1.y * v2.z - v1.z * v2.y;
		dst.backend.y = v1.z * v2.x - v1.x * v2.z;
		dst.backend.z = v1.x * v2.y - v1.y * v2.x;
		
		return dst;
	}
	
	div(factor: number): ae.math.Vector3D {
		
		this.backend.x /= factor;
		this.backend.y /= factor;
		this.backend.z /= factor;
		
		return this;
	}

	dot(v: ae.math.Vector3D): number {
		
		return (
			this.backend.x * v.backend.x + this.backend.y * v.backend.y +
			this.backend.z * v.backend.z);
	}
	
	getData(
			dst:    (ae.math.VectorBackend|ae.math.Vector3D|Array<number>),
			offset: number = 0):
			(ae.math.VectorBackend|ae.math.Vector3D|Array<number>) {
		
		if(dst instanceof Vector3D) {
			
			dst.setData(this.backend);
			
		} else if(dst instanceof ae.math.VectorBackend) {
			
			dst.x = this.backend.x;
			dst.y = this.backend.y;
			dst.z = this.backend.z;
			
		} else {
			
			dst[offset + 0] = this.backend.x;
			dst[offset + 1] = this.backend.y;
			dst[offset + 2] = this.backend.z;
		}
		
		return dst;
	}
	
	mult(factorOrV: (number|ae.math.Vector3D)): ae.math.Vector3D {
		
		if(factorOrV instanceof Vector3D) {
			
			this.backend.x *= factorOrV.backend.x;
			this.backend.y *= factorOrV.backend.y;
			this.backend.z *= factorOrV.backend.z;
			
		} else {
			
			this.backend.x *= factorOrV;
			this.backend.y *= factorOrV;
			this.backend.z *= factorOrV;
		}
		
		return this;
	}
	
	setData(
			xOrSrc: (ae.math.VectorBackend|ae.math.Vector3D|Array<number>|number),
			yOrOffset: number = 0,
			z:         number = 0): ae.math.Vector3D {
		
		if(xOrSrc instanceof Vector3D) {
			
			this.setData(this.backend);
			
		} else if(xOrSrc instanceof ae.math.VectorBackend) {
			
			this.backend.x = xOrSrc.x;
			this.backend.y = xOrSrc.y;
			this.backend.z = xOrSrc.z;
			
		} else if(typeof(xOrSrc) === "number") {
			
			this.backend.x = xOrSrc;
			this.backend.y = yOrOffset;
			this.backend.z = z;
			
		} else {
			
			this.backend.x = xOrSrc[yOrOffset + 0];
			this.backend.y = xOrSrc[yOrOffset + 1];
			this.backend.z = xOrSrc[yOrOffset + 2];
		}
		
		return this;
	}
	
	sub(v: ae.math.Vector3D): ae.math.Vector3D {
		
		this.backend.x -= v.backend.x;
		this.backend.y -= v.backend.y;
		this.backend.z -= v.backend.z;
		
		return this;
	}
	
	toZeroVector(): ae.math.Vector3D {
		
		return this.setData(0, 0, 0);
	}
};

File: lib/java/src/ae/math/Vector3D.java
package ae.math;

import ae.util.OrganizedObject;

public final class Vector3D extends OrganizedObject<Vector3D> {
	
	public final VectorBackend backend;
	public final Vector3D      readOnly;
	
	public float x, y, z;

	public Vector3D(final ReadOnlyBackend backend) {
		
		this.backend  = backend;
		this.readOnly = this;
		
		backend.addListener((obj) -> _propagateChange());
	}

	public Vector3D(final VectorBackend backend) {
		
		this.backend  = backend;
		this.readOnly = new Vector3D(new ReadOnlyBackend(backend));
		
		backend.addListener((obj) -> _propagateChange());
	}

	public final Vector3D add(final float value) {
		
		backend.setX(backend.getX() + value);
		backend.setY(backend.getY() + value);
		backend.setZ(backend.getZ() + value);
		
		return this;
	}

	public final Vector3D add(final Vector3D v) {
		
		backend.setX(backend.getX() + v.backend.getX());
		backend.setY(backend.getY() + v.backend.getY());
		backend.setZ(backend.getZ() + v.backend.getZ());
		
		return this;
	}

	public static final float angleDeg(
			final Vector3D v1,
			final Vector3D v2) {
		
		return (float)Math.toDegrees(angleRad(v1, v2));
	}
	
	public static final float angleRad(
			final Vector3D v1,
			final Vector3D v2) {
		
		return (float)
			Math.acos(dot(v1, v2) / (v1.computeLength() * v2.computeLength()));
	}
	
	public final Vector3D cloneConst() {
		return new Vector3D(new ReadOnlyBackend(new StaticBackend(backend)));
	}
	
	public final Vector3D cloneDeep() {
		return new Vector3D(backend);
	}
	
	public final Vector3D cloneStatic() {
		return new Vector3D(new StaticBackend(backend));
	}
	
	public final Vector3D copyStaticValues() {
		
		x = backend.getX();
		y = backend.getY();
		z = backend.getZ();
		
		return this;
	}
	
	public final float computeLength() {
		copyStaticValues();
		return (float)Math.sqrt(x * x + y * y + z * z);
	}

	public final float computeMean() {
		return (backend.getX() + backend.getY() + backend.getZ()) / 4;
	}

	public static final Vector3D createConst() {
		return createConst(0);
	}

	public static final Vector3D createConst(final float grey) {
		return createConst(grey, grey, grey);
	}
	
	public static final Vector3D createConst(
			final float x,
			final float y,
			final float z) {
		
		return new Vector3D(new ReadOnlyBackend(new StaticBackend(x, y, z, 1)));
	}

	public static final Vector3D createStatic() {
		return createStatic(0);
	}

	public static final Vector3D createStatic(final float grey) {
		return createStatic(grey, grey, grey);
	}
	
	public static final Vector3D createStatic(
			final float x,
			final float y,
			final float z) {
		
		return new Vector3D(new StaticBackend(x, y, z, 1));
	}
	
	public static final Vector3D cross(
			final Vector3D v1,
			final Vector3D v2,
			final Vector3D dst) {
		
		v1.copyStaticValues();
		v2.copyStaticValues();
		
		dst.backend.setX(v1.y * v2.z - v1.z * v2.y);
		dst.backend.setY(v1.z * v2.x - v1.x * v2.z);
		dst.backend.setZ(v1.x * v2.y - v1.y * v2.x);
		
		return dst;
	}
	
	public final Vector3D div(final float factor) {
		
		backend.setX(backend.getX() / factor);
		backend.setY(backend.getY() / factor);
		backend.setZ(backend.getZ() / factor);
		
		return this;
	}

	public final Vector3D div(final Vector3D v) {
		
		backend.setX(backend.getX() / v.backend.getX());
		backend.setY(backend.getY() / v.backend.getY());
		backend.setZ(backend.getZ() / v.backend.getZ());
		
		return this;
	}
	
	public static final float dot(
			final Vector3D v1,
			final Vector3D v2) {
		
		return
			v1.backend.getX() * v2.backend.getX() +
			v1.backend.getY() * v2.backend.getY() +
			v1.backend.getZ() * v2.backend.getZ();
	}
	
	public final VectorBackend getData(final VectorBackend dst) {
		
		dst.setX(backend.getX());
		dst.setY(backend.getY());
		dst.setZ(backend.getZ());
		
		return dst;
	}
	
	public final Vector3D getData(final Vector3D dst) {
		return dst.setData(backend);
	}
	
	public final float[] getData(final float[] dst) {
		
		dst[0] = backend.getX();
		dst[1] = backend.getY();
		dst[2] = backend.getZ();
		
		return dst;
	}
	
	public final float[] getData(
			final float[] dst,
			final int     offset) {
		
		dst[offset    ] = backend.getX();
		dst[offset + 1] = backend.getY();
		dst[offset + 2] = backend.getZ();
		
		return dst;
	}

	public final Vector3D mult(final float factor) {
		
		backend.setX(backend.getX() * factor);
		backend.setY(backend.getY() * factor);
		backend.setZ(backend.getZ() * factor);
		
		return this;
	}
	
	public final Vector3D mult(final Vector3D v) {
		
		backend.setX(backend.getX() * v.backend.getX());
		backend.setY(backend.getY() * v.backend.getY());
		backend.setZ(backend.getZ() * v.backend.getZ());
		
		return this;
	}
	
	public final Vector3D normalize() {
		return div(computeLength());
	}
	
	public final Vector3D setData(final VectorBackend src) {
		
		backend.setX(src.getX());
		backend.setY(src.getY());
		backend.setZ(src.getZ());
		
		return this;
	}
	
	public final Vector3D setData(final Vector3D src) {
		return setData(src.backend);
	}

	public final Vector3D setData(
			final float x,
			final float y,
			final float z) {
		
		backend.setX(x).setY(y).setZ(z);
		
		return this;
	}
	
	public final Vector3D setData(final float[] src) {
		return setData(src[0], src[1], src[2]);
	}
	
	public final Vector3D setData(
			final float[] src,
			final int     offset) {

		return setData(src[offset], src[offset + 1], src[offset + 2]);
	}

	public final Vector3D sub(final float value) {
		
		backend.setX(backend.getX() - value);
		backend.setY(backend.getY() - value);
		backend.setZ(backend.getZ() - value);
		
		return this;
	}

	public final Vector3D sub(final Vector3D v) {
		
		backend.setX(backend.getX() - v.backend.getX());
		backend.setY(backend.getY() - v.backend.getY());
		backend.setZ(backend.getZ() - v.backend.getZ());
		
		return this;
	}

	@Override
	public final String toString() {
		return toString(" ");
	}
	
	public final String toString(final String sep) {
		return toString("(", sep, ")");
	}
	
	public final String toString(
			final String start,
			final String sep,
			final String end) {
		
		return
			start +
			backend.getX() + sep + backend.getY() + sep + backend.getZ() +
			end;
	}
	
	public final Vector3D toZeroVector() {
		return setData(0, 0, 0);
	}
}
